src/internal/pfsdb/commit_provenance.go

```diff
@@ -166,3 +166,123 @@ func addCommitProvenance(tx *pachsql.Tx, from, to int) error {
 	_, err := tx.Exec(stmt, from, to)
 	return errors.Wrapf(err, "add commit provenance")
 }
+
+type commitWithDepth struct {
+	Commit
+	Depth uint32 `db:"depth"`
+}
+
+// GetCommitWithIDProvenance returns the full provenance of a branch, i.e. all branches that it either directly or transitively depends on.
+// It accepts a GraphOpt option.
+func GetCommitWithIDProvenance(ctx context.Context, ext sqlx.ExtContext, startId CommitID, opt ...GraphOpt) ([]*CommitWithID, error) {
+	commitsWithDepth, err := getCommitProvenanceWithDepth(ctx, ext, startId, opt...)
+	if err != nil {
+		return nil, errors.Wrap(err, "get commit with id provenance")
+	}
+	return commitWithDepthToCommitWithID(ctx, ext, commitsWithDepth)
+}
+
+func getCommitProvenanceWithDepth(ctx context.Context, ext sqlx.ExtContext, commitId CommitID, opt ...GraphOpt) ([]commitWithDepth, error) {
+	option := defaultOption()
+	if len(opt) > 0 {
+		option.Merge(opt[0])
+	}
+	var commits []commitWithDepth
+	if err := sqlx.SelectContext(ctx, ext, &commits, `
+		WITH RECURSIVE prov(from_id, to_id) AS (
+		    SELECT from_id, to_id, 1 as depth
+		    FROM pfs.commit_provenance
+		    WHERE from_id = $1
+		  UNION ALL
+		    SELECT cp.from_id, cp.to_id, depth+1
+		    FROM prov p
+		    JOIN pfs.commit_provenance cp ON cp.from_id = p.to_id
```

Comment:
Shouldn't there be a condition for the depth below this?

Reply:
You're right, can't believe I missed that